Update quick start guide (#108)
HubRise/website: content/developers/en/quick_start.md

  description:
---

## Overview

This guide will help you to get to know the HubRise API. At the end of the process, you will know how to send an order to HubRise. And the best part is that you will not have to write a single line of code!

## Getting Started

### Create a HubRise Account

First of all, you need to [create a HubRise account](https://www.hubrise.com/docs/getting-started/). It is free, and only takes a few minutes!
This account will be the first user of your application, and will be useful for testing during the development process.

### Set up Postman

Postman is an API development environment that makes sending API requests very easy. 

- Download and install [Postman](https://www.getpostman.com/).

> - Download the [HubRise API Postman Bundle]() and import the folder into Postman.

Author: Yes, having a single download button would be much better. We don't want to document how to download the bundle as well!! :)
However, the bundle will actually be two files in a folder, so the ideal workflow would be to put them in a zip file that the user downloads.

> - Download the [HubRise API Postman Bundle]() and import the folder into Postman.

Reviewer: Ok how about we generate the zip file, upload it to our Integrations shared folder "Integrations/HubRise documentation" and share a public link?

Also let's leave the postman collection in the `postman` repo for now. Can you also add a short `PUBLISH_ON_WEBSITE.md` file in the bundle folder there to explain the process in 2-3 sentences?

> - Download the [HubRise API Postman Bundle]() and import the folder into Postman.

Author: Should this process be automatic (eg, every time the collection is updated)?

> - Download the [HubRise API Postman Bundle]() and import the folder into Postman.

Reviewer: No need to, even a slightly old collection would still be very useful. Just make the process easy, by documenting it.

> - Download the [HubRise API Postman Bundle]() and import the folder into Postman.

Author: OK perfect! I've added the file in the meantime.

> - Download the [HubRise API Postman Bundle]() and import the folder into Postman.

Reviewer: Have you added it? I cannot see it (in the integrations folder). Also don't forget to update the documentation to include a link to the collection?

### Create a HubRise Account

> First of all, you need to create a user profile on HubRise with an account and an associated location. It is free, and only takes a few minutes! From the [HubRise home page](https://www.hubrise.com/), click on the **Sign Up** button and follow the instructions. 

Reviewer: No need to go to the homepage: the button is actually available on the very page they are reading

Author: Ups! You're right... fixed that!

}
```

The client's **id** and **secret** uniquely identify your application. You will generally need to create a single client for your application: The same client can be used by several HubRise accounts.

## Request the Account's Authorisation

In order for your application to access a HubRise account's information, you need the authorisation from the account's owner. 

You, as the owner of your test HubRise account, must authorise access to your own application.

Open a web browser with the URL and make sure you copy the `client_id` from the secret JSON you downloaded before in place of `THE-CLIENT-ID`:

```http
https://manager.hubrise.com/oauth2/v1/authorize?
  redirect_uri=urn:ietf:wg:oauth:2.0:oob&
  client_id=THE-CLIENT-ID&
  scope=location[orders.write,customer_list.write,catalog.read]
```

HubRise will then:
- Authenticate the user.

> - Ask to choose the location, account, catalog and customer list to connect.

Author: OK, I simplified the wording of this paragraph.

- Obtain consent to access the requested scope. 

If everything goes well, you should see a page similar to the following. 

![User code](../images/001-en-generate-user-code.png)

> This is the **user code**: copy it and head to the next section. 

Author: OK, changed it both in the doc and in the Postman collection

## Generate the Access Token

You now have three codes:

- The client id
- The client secret
- The user code

With these, you can generate your access token with Postman.

An access token is how you authenticate the requests you send to HubRise.
Basically, you want your application (_client id_) to be authorised (via the _client secret_) to send requests to HubRise on behalf of the account owner (_user code_), and you do this via an _access token_.

### Set up Variables in Postman

> In Postman, select the **HubRise API** environment and enter the values for `client_id`, `client_secret` and `user_code`. 

Author: Yes! If you do "Import folder", Postman will automatically detect the collection and its environment.

![Screenshot of Postman](../images/002-postman-environment.png)

### Send the Request

> Select the **Generate Access Token** request and click the **Send** button. 

Reviewer: I now realize that you are referring to your own Postman collection. What do you think of merging your collection into the one already in the `postman` repo?

Author: Yes, I've done that now.

Reviewer: Wow, great!

Postman will send the following request to the API:

```json
POST https://manager.hubrise.com/oauth2/v1/token?code={{code}}&client_id={{client_id}}&client_secret={{client_secret}}
```

If the request is successful, Postman will automatically create the `hubrise_access_token` variable in the environment.

All the following requests in Postman will use the `X-Access-Token` header with this access token.

## Create Your First Order

Now you are ready to send valid requests to HubRise and to create new orders.

In Postman, play the **Create Order** endpoint. If the request is successful, you should see a new order in the [HubRise back office](https://manager.hubrise.com/orders).

> ## Experiment with the API

Reviewer: To go further:
- the reader can install the free OrderLine tool to check how his orders look
- he can convert any Postman request into a snippet in his programming language, to transition from "experimenting" to "integrating"

Reviewer: Before "Experiment with the API", we should add another chapter "Setup a callback". Seems a bit of a stretch for a quick start guide, but under your guidance it will take a few minutes for the developer to complete this step and this is a topic we often have questions about.

Here are the steps:
- The developer needs to create a new client, download id/secret, and generate a new token. - On the first client, register a callback with URL pointing to a webhook.site page, listening to order.create events
- On the new client, create an order
- Check that is is received in webhook.site (active polling)
- If the developer cannot expose an URL (for instance it's a Windows installed app), add a note that he can also GET /events at regular intervals

Author: OK, I still need to add the webhook section.

Author: Hi @antmonnier! Regarding the callback, how about reversing the roles? So we register the callback on the second client, so that we can create the order with the first client? Otherwise the user would have to change the credentials in the Postman collection...

Reviewer: Hi @atorin As you like but the user will still need to send an API request from the 2nd client to register the callback URL.

Author: @antmonnier OK, I've added the callback section. Let me know if it works...

}
```

The client's **id** and **secret** uniquely identify your application. You will generally need to create a single client for your application: The same client can be used by several HubRise accounts.

## Request the Account's Authorisation

In order for your application to access a HubRise account's information, you need the authorisation from the account's owner. 

You, as the owner of your test HubRise account, must authorise access to your own application.

Open a web browser with the URL and make sure you copy the `client_id` from the secret JSON you downloaded before in place of `THE-CLIENT-ID`:

```http
https://manager.hubrise.com/oauth2/v1/authorize?
  redirect_uri=urn:ietf:wg:oauth:2.0:oob&

>   client_id=THE-CLIENT-ID&

Reviewer: You have used a different convention (`{{client_id}}`) a bit further below.

Author: @antmonnier Which one do you think it's better?

Author: I've used the brackets version for now, but we can change that in case.

### Set up Variables in Postman

In Postman, select the **HubRise API** environment and enter the values for `client_id`, `client_secret` and `user_code`. 

![Screenshot of Postman](../images/002-postman-environment.png)

### Send the Request

Select the **Generate Access Token** request and click the **Send** button. 

Postman will send the following request to the API:

```json
POST https://manager.hubrise.com/oauth2/v1/token?code={{code}}&client_id={{client_id}}&client_secret={{client_secret}}
```

If the request is successful, Postman will automatically create the `hubrise_access_token` variable in the environment.

All the following requests in Postman will use the `X-Access-Token` header with this access token.

## Create Your First Order

Now you are ready to send valid requests to HubRise and to create new orders.

In Postman, play the **Create Order** endpoint. If the request is successful, you should see a new order in the [HubRise back office](https://manager.hubrise.com/orders).

Congratulations! You placed your first order on HubRise.

## (Optional) Set up a Callback

> 

Reviewer: Maybe a short intro here:
"In many cases your application will need to be notified when an event such as an order creation or update occurs."

------

**IMPORTANT NOTE**: An application cannot receive events generated by itself!

------

### Obtain the Access Token

After creating the new client, head to the Postman collection and populate the `callback_client_id`, `callback_client_secret`, and `callback_auth_code` fields in the environment. 

You can use the `Callback: Generate Access Token` request to obtain the access token. This will be automatically saved as a separate environment variable from the previous one, since you will need both.

### Obtain the Webhook URL

In order to register your callback you need a valid URL that can listens to events. If you don't have your own server already, you might use the [webhook.site](https://webhook.site/) service. When you visit the website, a unique URL will be created that you can use as your webhook URL. 

This page will show all the events received by the callback.

### Register the Callback

Now that you have the callback URL, head back to Postman and copy the value in the `callback_url` field. Then, play the `Callback: Create Webhook` request. 

A successful response means that you registered your callback URL correctly.

### Place an Order and Check the Webhook

> Now you can try to place an order with your first application. You can use the same `Create Order` request you sent before. This time, however, you will see that the order is automatically forwarded to your webhook URL. 

Reviewer: Just to make sure users don't get stuck (that's such a common trap), would be worth mentioning one time again sth like: "The event should appear almost immediately on the webhook.site page. If it fails to appear, check that the order was placed using the token from your first application.".

> ### Get New Events without Callback

Reviewer: It's technically still a callback, just a passive one. Maybe change the title to "Get New Events without a Public URL"?

> In some circumstances, a webhook cannot be set up because a callback URL cannot be exposed. This is typical for applications that run on Windows.

Reviewer: ```suggestion
In some circumstances, a webhook cannot be set up because a callback URL cannot be exposed. This is typical for applications that users download and run from their computer.
```

In this case, you can set up a _passive callback_: your application can check for new events at regular intervals of time and perform the necessary operations when a new event is found. 

> The `Callback: Get Events` request in Postman is an example of how you can do that. Simply create a new order, as usual, and play this request right after to obtain the order's details.

Reviewer: Unlike active callbacks, events need to be manually deleted after being processed.

## Next Steps

### Install OrderLine

OrderLine is a simple, free application that you can connect to your HubRise account to manage incoming orders. With OrderLine, you will be able to accept or reject incoming test orders and to inspect the different logs from Postman.

To connect OrderLine to your location, click on **CONNECTIONS > View available apps**, then scroll down the list and select OrderLine.

### Experiment with the API

The Postman collection contains other typical requests to obtain the account's information, create and update customers, and update order statuses. Test the various endpoints according to your needs.

### From Postman to Code

Postman has a useful built-in functionality that allows to transform your test requests to actual code snippets in various languages. In this way, you will be able to transition from testing to integrating really quickly.

### Explore the API

> To learn more about the HubRise API, you can explore our [API Reference](../api/general-concepts), or you can visit the [Authentication](../authentication) page to learn more about the possible authentication mechanisms.

Reviewer: URLs open in a new tab by default, unless they start with a slash `/`. Internal links should avoid opening a new tab:

```suggestion
To learn more about the HubRise API, you can explore our [API Reference](/developers/api/general-concepts), or you can visit the [Authentication](/developers/authentication/) page to learn more about the possible authentication mechanisms.
```

Author: OK thanks! I'll change that
